File: Sources/custom_adapter/TopOnAdapter/Objc/6.5.x/Splash/AlxTopOnSplashAdapter.m
//
//  AlxTopOnSplashAdapter.m
//  AlxAdsOCDemo
//

#import "AlxTopOnSplashAdapter.h"
#import "AlxTopOnBaseManager.h"
#import "AlxTopOnSplashDelegate.h"
#import "AlxTopOnSplashEvent.h"
#import "AlxTopOnBiddingRequest.h"
#import "AlxTopOnBiddingRequestManager.h"
#import "AlxTopOnTool.h"

static NSString *const TAG = @"AlxTopOnSplashAdapter";

@interface AlxTopOnSplashAdapter ()

@property (nonatomic, strong) AlxSplashAd *splashAd;
@property (nonatomic, strong) AlxTopOnSplashDelegate *splashDelegate;

@end

@implementation AlxTopOnSplashAdapter

@synthesize adStatusBridge = _adStatusBridge;

#pragma mark - Lazy Load

- (AlxTopOnSplashDelegate *)getSplashDelegate {
    if (!_splashDelegate) {
        _splashDelegate = [[AlxTopOnSplashDelegate alloc] init];
        _splashDelegate.adStatusBridge = self.adStatusBridge;
    }
    return _splashDelegate;
}

#pragma mark - ATBaseSplashAdapterProtocol Ad Load

- (void)loadADWithArgument:(ATAdMediationArgument *)argument {
    NSLog(@"%@: loadAD", TAG);
    NSLog(@"%@: loadAD: isMainThread=%@", TAG, [NSThread isMainThread] ? @"YES" : @"NO");

    NSString *bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey];
    NSLog(@"%@: loadAD: bidId=%@", TAG, bidId ?: @"空");

    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *unitKeyStr = [AlxTopOnBaseManager unitID];
        NSString *unitId = argument.serverContentDic[unitKeyStr];

        if (!unitId || unitId.length == 0) {
            NSString *errorStr = @"unitid is empty";
            NSLog(@"%@: loadAD: error = %@", TAG, errorStr);
            NSError *error = [AlxTopOnBaseManager errorWithCode:-100 message:errorStr];
            [self notifyLoadFailed:error];
            return;
        }
        NSLog(@"%@: loadAD: unitid = %@", TAG, unitId);

        // 自定义底部视图
        UIView *bottomView = nil;
        if ([argument.localInfoDic[kATSplashExtraNewBottomViewKey] isKindOfClass:[UIView class]]) {
            bottomView = argument.localInfoDic[kATSplashExtraNewBottomViewKey];
        } else if ([argument.serverContentDic[kATSplashExtraNewBottomViewKey] isKindOfClass:[UIView class]]) {
            bottomView = argument.serverContentDic[kATSplashExtraNewBottomViewKey];
        } else if ([argument.localInfoDic[kATSplashExtraContainerViewKey] isKindOfClass:[UIView class]]) {
            bottomView = argument.localInfoDic[kATSplashExtraContainerViewKey];
        }

        // autoCloseOnFinish
        BOOL autoClose = NO;
        if (argument.localInfoDic[@"autoCloseOnFinish"]) {
            autoClose = [argument.localInfoDic[@"autoCloseOnFinish"] boolValue];
        } else if (argument.serverContentDic[@"autoCloseOnFinish"]) {
            autoClose = [argument.serverContentDic[@"autoCloseOnFinish"] boolValue];
        } else if (argument.localInfoDic[@"auto_close"]) {
            autoClose = [argument.localInfoDic[@"auto_close"] boolValue];
        } else if (argument.serverContentDic[@"auto_close"]) {
            autoClose = [argument.serverContentDic[@"auto_close"] boolValue];
        } else if (argument.localInfoDic[@"is_auto_close"]) {
            autoClose = [argument.localInfoDic[@"is_auto_close"] boolValue];
        } else if (argument.serverContentDic[@"is_auto_close"]) {
            autoClose = [argument.serverContentDic[@"is_auto_close"] boolValue];
        }

        if (bidId) {
            // Bidding 场景
            AlxTopOnBiddingRequest *biddingRequest = (AlxTopOnBiddingRequest *)[[AlxTopOnTool shared] getRequestItemWithUnitID:unitId];
            if (biddingRequest) {
                self.splashAd = (AlxSplashAd *)biddingRequest.customObject;
                if (bottomView) {
                    self.splashAd.customBottomView = bottomView;
                }
                self.splashAd.autoCloseOnFinish = autoClose;

                if (self.splashAd) {
                    NSLog(@"%@: loadAD: bid ad loaded, notify success", TAG);
                    NSMutableDictionary *adExtra = [NSMutableDictionary dictionary];
                    adExtra[kATAdAssetsCustomObjectKey] = self.splashAd;
                    [self notifySplashLoaded:adExtra];
                } else {
                    NSLog(@"%@: loadAD: bid ad object is empty", TAG);
                    NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid ad object is empty"}];
                    [self notifyLoadFailed:error];
                }
            } else {
                NSLog(@"%@: loadAD: bid request not found in cache", TAG);
                NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid request not found"}];
                [self notifyLoadFailed:error];
            }
            [[AlxTopOnTool shared] removeRequestItemWithUnitID:unitId];
        } else {
            // 普通加载场景
            self.splashAd = [[AlxSplashAd alloc] init];
            self.splashAd.delegate = [self getSplashDelegate];
            self.splashDelegate.splashAd = self.splashAd;
            self.splashAd.customBottomView = bottomView;
            self.splashAd.autoCloseOnFinish = autoClose;

            NSLog(@"%@: start loading splash ad with unitId: %@", TAG, unitId);
            AlxAdRequest *req = [[[AlxAdRequest alloc] init] withUserExt:@{ @"bid_floor": @"1.68" }];
            [self.splashAd loadAdWithAdUnitId:unitId request:req];
        }
    });
}

#pragma mark - Helper Notifications

- (void)notifySplashLoaded:(NSDictionary *)adExtra {
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnSplashAdLoadedExtra:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:adExtra];
#pragma clang diagnostic pop
        }
    }
}

- (void)notifyLoadFailed:(NSError *)error {
    if (self.adStatusBridge) {
        SEL selector = NSSelectorFromString(@"atOnAdLoadFailed:adExtra:");
        if ([self.adStatusBridge respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            [self.adStatusBridge performSelector:selector withObject:error withObject:@{}];
#pragma clang diagnostic pop
        }
    }
}

#pragma mark - C2S Header Bidding

+ (void)bidRequestWithPlacementModel:(ATPlacementModel *)placementModel
                      unitGroupModel:(ATUnitGroupModel *)unitGroupModel
                                info:(NSDictionary *)info
                          completion:(void (^)(ATBidInfo * _Nullable, NSError * _Nullable))completion {
    NSLog(@"%@: bidRequestWithPlacementModel", TAG);

    dispatch_async(dispatch_get_main_queue(), ^{
        if (![AlxTopOnBaseManager isInitialized]) {
            [AlxTopOnBaseManager initSDKWithServerInfo:info];
        }
    });

    AlxTopOnSplashEvent *customEvent = [[AlxTopOnSplashEvent alloc] initWithInfo:info localInfo:info];
    customEvent.isC2SBiding = YES;

    AlxTopOnBiddingRequest *request = [[AlxTopOnBiddingRequest alloc]
                                       initWithUnitGroup:unitGroupModel
                                       customEvent:customEvent
                                       unitID:info[[AlxTopOnBaseManager unitID]]
                                       placementID:placementModel.placementID
                                       extraInfo:info
                                       adType:ATAdFormatSplash
                                       bidCompletion:completion];

    [[AlxTopOnBiddingRequestManager shared] startWithRequest:request];
}

#pragma mark - Ad Ready Check

- (BOOL)adReadySplashWithInfo:(NSDictionary *)info {
    NSLog(@"%@: adReadySplashWithInfo", TAG);
    if (self.splashAd && [self.splashAd isReady]) {
        NSLog(@"%@: adReady = YES", TAG);
        return YES;
    }
    NSLog(@"%@: adReady = NO", TAG);
    return NO;
}

+ (BOOL)adReadyWithCustomObject:(id)customObject info:(NSDictionary *)info {
    if ([customObject isKindOfClass:[AlxSplashAd class]]) {
        return [(AlxSplashAd *)customObject isReady];
    }
    return NO;
}

#pragma mark - Show Splash Ad

- (void)showSplashAdInWindow:(UIWindow *)window inViewController:(UIViewController *)inViewController parameter:(NSDictionary *)parameter {
    NSLog(@"%@: showSplashAdInWindow:inViewController:parameter:", TAG);
    dispatch_async(dispatch_get_main_queue(), ^{
        if (!self.splashAd) {
            NSLog(@"%@: splashAd is nil", TAG);
            return;
        }

        if (parameter) {
            if ([parameter[kATSplashExtraNewBottomViewKey] isKindOfClass:[UIView class]]) {
                self.splashAd.customBottomView = parameter[kATSplashExtraNewBottomViewKey];
            } else if ([parameter[kATSplashExtraContainerViewKey] isKindOfClass:[UIView class]]) {
                self.splashAd.customBottomView = parameter[kATSplashExtraContainerViewKey];
            }

            if (parameter[@"autoCloseOnFinish"]) {
                self.splashAd.autoCloseOnFinish = [parameter[@"autoCloseOnFinish"] boolValue];
            } else if (parameter[@"auto_close"]) {
                self.splashAd.autoCloseOnFinish = [parameter[@"auto_close"] boolValue];
            } else if (parameter[@"is_auto_close"]) {
                self.splashAd.autoCloseOnFinish = [parameter[@"is_auto_close"] boolValue];
            }
        }

        [self.splashAd showAdInWindow:window];
    });
}

@end

File: Sources/custom_adapter/TopOnAdapter/Swift/6.5.x/Splash/AlxTopOnSplashAdapter.swift
//
//  AlxTopOnSplashAdapter.swift
//  AlxAdsDemo
//
//  TopOn 6.5.x 新架构
//

import Foundation
import UIKit
import AnyThinkSDK
import AlxAds

@objc(AlxTopOnSplashAdapter)
public class AlxTopOnSplashAdapter: AlxTopOnBaseAdapter, ATBaseSplashAdapterProtocol {
    
    private static let TAG = "AlxTopOnSplashAdapter"
    
    @objc public var adStatusBridge: ATSplashAdStatusBridge!
    
    private var splashAd: AlxSplashAd?
    private var splashDelegate: AlxTopOnSplashDelegate?
    
    // MARK: - 懒加载 / Lazy Load
    private func getSplashDelegate() -> AlxTopOnSplashDelegate {
        if splashDelegate == nil {
            splashDelegate = AlxTopOnSplashDelegate()
            splashDelegate?.adStatusBridge = self.adStatusBridge
        }
        return splashDelegate!
    }
    
    // MARK: - 广告加载 / Ad Load (ATBaseSplashAdapterProtocol)
    @objc public override func loadAD(with argument: ATAdMediationArgument) {
        NSLog("%@: loadAD", AlxTopOnSplashAdapter.TAG)
        NSLog("%@: loadAD: isMainThread=%@", AlxTopOnSplashAdapter.TAG, Thread.current.isMainThread ? "YES" : "NO")
        
        let bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey] as? String
        NSLog("%@: loadAD: bidId=%@", AlxTopOnSplashAdapter.TAG, bidId ?? "空")
        
        DispatchQueue.main.async {
            let unitIdKey = AlxTopOnBaseManager.unitID
            #if DEBUG
            let unitId = "203688"
            #else
            guard let unitId = argument.serverContentDic[unitIdKey] as? String, !unitId.isEmpty else {
                let errorStr = "unitid is empty"
                NSLog("%@: loadAD: error = %@", AlxTopOnSplashAdapter.TAG, errorStr)
                let error = AlxTopOnBaseManager.error(code: -100, msg: errorStr)
                self.notifyLoadFailed(error: error)
                return
            }
            #endif
            NSLog("%@: loadAD: unitid = %@", AlxTopOnSplashAdapter.TAG, unitId)
            
            // 提取自定义底部视图 (若有)
            var bottomView: UIView? = nil
            if let customBottom = argument.localInfoDic[kATSplashExtraNewBottomViewKey] as? UIView {
                bottomView = customBottom
            } else if let customBottom = argument.serverContentDic[kATSplashExtraNewBottomViewKey] as? UIView {
                bottomView = customBottom
            } else if let customBottom = argument.localInfoDic[kATSplashExtraContainerViewKey] as? UIView {
                bottomView = customBottom
            }
            
            // 提取 autoCloseOnFinish 配置 (若有，默认 false)
            var autoClose: Bool = false
            if let v = argument.localInfoDic["autoCloseOnFinish"] as? Bool {
                autoClose = v
            } else if let v = argument.serverContentDic["autoCloseOnFinish"] as? Bool {
                autoClose = v
            } else if let v = argument.localInfoDic["auto_close"] as? Bool {
                autoClose = v
            } else if let v = argument.serverContentDic["auto_close"] as? Bool {
                autoClose = v
            } else if let v = argument.localInfoDic["is_auto_close"] as? Bool {
                autoClose = v
            } else if let v = argument.serverContentDic["is_auto_close"] as? Bool {
                autoClose = v
            }
            
            if bidId != nil {
                // Bidding 场景：从缓存中取出已加载的广告
                if let biddingRequest = AlxTopOnTool.shared.getRequestItem(withUnitID: unitId) as? AlxTopOnBiddingRequest {
                    self.splashAd = biddingRequest.customObject as? AlxSplashAd
                    if let bottomView = bottomView {
                        self.splashAd?.customBottomView = bottomView
                    }
                    self.splashAd?.autoCloseOnFinish = autoClose
                    
                    if let splashAd = self.splashAd {
                        NSLog("%@: loadAD: bid ad loaded, notify success", AlxTopOnSplashAdapter.TAG)
                        var adExtra: [AnyHashable: Any] = [:]
                        adExtra[kATAdAssetsCustomObjectKey] = splashAd
                        self.notifySplashLoaded(adExtra: adExtra)
                    } else {
                        NSLog("%@: loadAD: bid ad object is empty", AlxTopOnSplashAdapter.TAG)
                        let error = NSError(domain: "AlxTopOnAdapter", code: -100, userInfo: [NSLocalizedDescriptionKey: "Bid ad object is empty"])
                        self.notifyLoadFailed(error: error)
                    }
                } else {
                    NSLog("%@: loadAD: bid request not found in cache", AlxTopOnSplashAdapter.TAG)
                    let error = NSError(domain: "AlxTopOnAdapter", code: -100, userInfo: [NSLocalizedDescriptionKey: "Bid request not found"])
                    self.notifyLoadFailed(error: error)
                }
                AlxTopOnTool.shared.removeRequestItem(withUnitID: unitId)
            } else {
                // 普通加载场景
                self.splashAd = AlxSplashAd()
                self.splashAd?.delegate = self.getSplashDelegate()
                self.splashDelegate?.splashAd = self.splashAd
                self.splashAd?.customBottomView = bottomView
                self.splashAd?.autoCloseOnFinish = autoClose
                
                NSLog("%@: start loading splash ad with unitId: %@", AlxTopOnSplashAdapter.TAG, unitId)
                let req = AlxAdRequest().withUserExt([
                    "bid_floor": "1.68"
                ])
                self.splashAd?.loadAd(adUnitId: unitId, request: req)
            }
        }
    }
    
    // MARK: - 辅助通知方法
    private func notifySplashLoaded(adExtra: [AnyHashable: Any]) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnSplashAdLoadedExtra:")
            if bridge.responds(to: selector) {
                bridge.perform(selector, with: adExtra)
            }
        }
    }
    
    private func notifyLoadFailed(error: Error) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnAdLoadFailed:adExtra:")
            if bridge.responds(to: selector) {
                bridge.perform(selector, with: error, with: nil)
            }
        }
    }
    
    // MARK: - C2S Header Bidding 竞价
    @objc public static func bidRequestWithPlacementModel(_ placementModel: ATPlacementModel,
                                                          unitGroupModel: ATUnitGroupModel,
                                                          info: [AnyHashable: Any],
                                                          completion: @escaping (ATBidInfo?, Error?) -> Void) {
        NSLog("%@: bidRequestWithPlacementModel", AlxTopOnSplashAdapter.TAG)
        
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: info)
            }
        }
        
        let customEvent = AlxTopOnSplashEvent(info: info, localInfo: info)
        customEvent.isC2SBiding = true
        
        let request = AlxTopOnBiddingRequest(
            unitGroup: unitGroupModel,
            customEvent: customEvent,
            unitID: info[AlxTopOnBaseManager.unitID] as? String,
            placementID: placementModel.placementID,
            extraInfo: info,
            adType: ATAdFormat.splash,
            bidCompletion: completion
        )
        AlxTopOnBiddingRequestManager.shared.start(with: request)
    }
    
    // MARK: - 广告就绪检查
    @objc public func adReadySplash(withInfo info: [AnyHashable: Any]) -> Bool {
        NSLog("%@: adReadySplashWithInfo", AlxTopOnSplashAdapter.TAG)
        if let ad = self.splashAd, ad.isReady() {
            NSLog("%@: adReady = YES", AlxTopOnSplashAdapter.TAG)
            return true
        }
        NSLog("%@: adReady = NO", AlxTopOnSplashAdapter.TAG)
        return false
    }
    
    @objc public static func adReady(withCustomObject customObject: Any, info: [AnyHashable: Any]) -> Bool {
        if let splashAd = customObject as? AlxSplashAd {
            return splashAd.isReady()
        }
        return false
    }
    
    // MARK: - 广告展示 (ATBaseSplashAdapterProtocol)
    @objc(showSplashAdInWindow:inViewController:parameter:)
    public func showSplashAd(in window: UIWindow, in inViewController: UIViewController, parameter: [AnyHashable: Any]?) {
        NSLog("%@: showSplashAdInWindow:inViewController:parameter:", AlxTopOnSplashAdapter.TAG)
        DispatchQueue.main.async {
            guard let splashAd = self.splashAd else {
                NSLog("%@: splashAd is nil", AlxTopOnSplashAdapter.TAG)
                return
            }
            
            // 检查是否有自定义底部视图
            if let param = parameter {
                if let customBottom = param[kATSplashExtraNewBottomViewKey] as? UIView {
                    splashAd.customBottomView = customBottom
                } else if let customBottom = param[kATSplashExtraContainerViewKey] as? UIView {
                    splashAd.customBottomView = customBottom
                }
                
                if let v = param["autoCloseOnFinish"] as? Bool {
                    splashAd.autoCloseOnFinish = v
                } else if let v = param["auto_close"] as? Bool {
                    splashAd.autoCloseOnFinish = v
                } else if let v = param["is_auto_close"] as? Bool {
                    splashAd.autoCloseOnFinish = v
                }
            }
            
            splashAd.showAd(inWindow: window)
        }
    }
}
